CI, TST: Run Cygwin CI with Netlib reference BLAS and re-enable linalg tests

DWesl opened the pull request:
The new OpenBLAS version seems to be causing problems with the linalg tests.  This checks that hypothesis by removing the OpenBLAS optimized BLAS, leaving only the Netlib reference BLAS.

This also removes the `xfail` marks on the failing tests.
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

DWesl committed:
CI: Run Cygwin CI with Netlib reference BLAS.

mattip commented:
Tests pass:
```
17945 passed, 268 skipped, 1293 deselected, 19 xfailed, 5 xpassed, 1 warning in 344.24s (0:05:44)
```

Why does the cygwin CI run [install libopenblas](https://github.com/numpy/numpy/runs/4621657479?check_suite_focus=true#step:3:428) from cygwin rather than use the one from https://github.com/MacPython/openblas-libs downloaded with the tools/openblas_support.py script?

mattip commented:
Can we get some eyes on this from the people in cygwin who are building libopenblas?

DWesl commented:
Probably the biggest reason to use the Cygwin OpenBLAS over the Windows one is that [the type sizes differ; `long int` is 64 bits on 64-bit Cygwin compared to 32 bits on 64-bit Windows](https://cygwin.com/faq/faq.html#faq.programming.64bitporting).  Cygwin also assumes that most programs it calls will be Cygwin programs, so [it doesn't bother updating the Windows environment unless you ask it to](https://cygwin.com/cygwin-ug-net/setup-env.html#setup-env-win32).  Cygwin also provides its own C Runtime Library (equivalent to linux's `-lc` or [Windows's MSVCRT](https://cygwin.com/faq/faq.html#faq.programming.msvcrt-and-cygwin)), so using a DLL that makes extensive use of C library functions could cause confusion for users who expect behavior from one CRT and get behavior from another.

However, if someone is careful, they could [write a Cygwin wrapper around Windows OpenBLAS](https://cygwin.com/faq/faq.html#faq.programming.linking-lib).  I *think* Cygwin and Windows use the same function call conventions, and if not there's a way to declare functions that use the other function call convention.

Alternately, I could grab Cygwin's source package for OpenBLAS, pull out the build process, and put that into a CI run in the MacPython/openblas-libs so there's a download available for Cygwin.

If you're asking why Cygwin OpenBLAS provides only the standard BLAS interface, I think it's so that people can take a program linked against the reference BLAS and have it use OpenBLAS instead by installing a single package and using some `PATH` tricks (equivalent to `LD_LIBRARY_PATH` tricks on linux), or do what I do in this PR and force something to use the Netlib reference BLAS rather than the OpenBLAS optimized BLAS with only a single line of code.

The 3.10 run fails in `actions/checkout`, which seems unrelated.  

Since using the Netlib reference BLAS instead of the OpenBLAS optimized BLAS makes the tests pass, I suspect the Cygwin CI test failures were due to changes between OpenBLAS 0.3.18 and OpenBLAS 0.3.19; PR #20660 is to check whether these problems recur with the most recent OpenBLAS on non-Cygwin platforms.

charris commented:
@DWesl Looks like cygwin on main is passing again. This PR also looks to be passing. Do you want to keep it, or can it be closed?

DWesl commented:
I got the Cygwin tests on main passing by marking all the failing tests as xfail.  Some people expressed concern about just ignoring those tests, so I created this PR to make those tests passing part of the Cygwin check again, and to confirm a guess about whether the failures are related to the BLAS implementation.  #20660 is another approach to the second part, but I suspect making those tests pass on Cygwin still requires the first bit.

charris commented:
> I got the Cygwin tests on main passing by marking all the failing tests as xfail.

Ah, OK. This passes, so I guess that puts the burden on OpenBLAS or Cygwin.

DWesl commented:
> Ah, OK. This passes, so I guess that puts the burden on OpenBLAS or Cygwin.

Given that #20660 was failing on runners with AVX512, I suspect this is an OpenBLAS thing.

seberg commented:
I would be happy to juts merge this, ideally with one more comment as to why so that we can remove it easier when/if it becomes irrelevant.

My opinion here would be, that it is better that people know what is failing.  @DWesl do you still want this merged, then we will just put it in.

DWesl committed:
CI: Add explanatory comment to workaround step

This should function as a reminder to revert the change (and start using OpenBLAS again) the next time Cygwin OpenBLAS updates.

DWesl commented:
> I would be happy to juts merge this, ideally with one more comment as to why so that we can remove it easier when/if it becomes irrelevant.
> 
Does the comment in the most recent commit work?

seberg commented:
Thanks, perfect.  Cygwin tests are passing, so lets get this in.

DWesl commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -539,11 +534,6 @@ def test_types(self, dtype):
 
-    @pytest.mark.xfail(sys.platform == 'cygwin',
-                       reason="Consistently fails on CI.")
-    def test_sq_cases(self):
```
These are marking tests broken by OpenBLAS 0.3.19, but since this PR forces the CI to use netlib reference BLAS, those tests no longer fail.  If you want to keep them marked for local tests run with OpenBLAS, I can make another PR to deal with that, which would basically be a copy of #20654.

seberg replied:
It looks like the `def test_...` also got deleted, that is what I missed?

DWesl replied:
The body of the test just calls the implementation from the superclass; if there is no mark on the function definition, then there is no need to have a function that manually does what inheritance already does.

The mark was there because these tests fail on AVX-512 with OpenBLAS 0.3.19, since I wanted to mark the test as `xfail` only in the case where it actually failed, I had to define a function with the same name in the class where it failed, then apply the decorator to the new function.  Calling the superclass implementation seemed to be the simplest method of keeping everything straight, and verifying that this doesn't actually change the behavior of the class.
